File: mTurk/pilot-jspsych-experiment/conditions/compLists.py
#!/usr/bin/python
import json
import cgi
import os
import sys
import time
from datetime import datetime
sys.stderr = sys.stdout # Make sure we can see any errors

try:
  # Read the id
  fs = cgi.FieldStorage()
  amtId = fs["id"].value
  if fs["experimentName"].value != 'pilot':
    raise Exception('An experimentName value was loaded but did not match')
  
  # Open participants file to check time limit 
  pFile = open("participants.txt", "r")
  pString = pFile.read()
  pList = json.loads(pString)

  sFile = open("status.txt", "r")
  sString = sFile.read()
  sList = json.loads(sString)

  # Close the file and tell jQuery all went well:
  sFile.close()
  pFile.close()

  # Loop through list and check if over time
  for i, j in enumerate(pList):
    if pList[i][0] == amtId:
      index = i
      # Update status listings
      sList[index] = 2
      break

  # Now write the updated list into JSON
  with open("status.txt", "w") as sUpdate:
    json.dump(sList, sUpdate)
  
  # Close the file and tell jQuery all went well:
  pUpdate.close()
  sUpdate.close()

  print "Status: 200 OK"
  print "Content-type: text/plain"
  print
  print datetime.now().strftime('%Y-%m-%d %H:%M:%S') + " status.txt and participants.txt updated with a sequence completion"
except:

  # Tell jQuery something went wrong
  print "Status: 400 Bad Request"
  print "Content-type: text/plain"
  print
  print "Error"

  # Utilize participants.txt for returning users (refresh or otherwise) to deny
  # completion and ask for return

  # Create status.txt and participants.txt for JSON format
  # status.txt with be formed with [e] * n with e = 0 and n = number of
  # participants
  # participants.txt will be formed with [['p','t']] * n

  # Read status.txt into list
  # Check participants.txt (containing ID's and times accepted)
  # If there are times beyond the set expiration limit, find the associated ID 
  # (it may be okay to just have the index associated with file name 
  # rather than the ID itself) and adjust the status of the location back to 
  # "open"

File: mTurk/pilot-jspsych-experiment/conditions/compLists.pl
#!/usr/bin/perl -l
use CGI qw(:standard);
use CGI::Carp qw(warningsToBrowser fatalsToBrowser);  # Make sure we can see any errors
use lib;
use JSON;

print header;
# Read the id
my $cgi = CGI->new;

# Parse out the three fields we know the experiments always send us,
# the id, the experimentName, and the data (curData). Save the data
# in a file called {id}.txt in folder {experimentName}.
my $id = $cgi->param('id');
my $experimentName = $cgi->param('experimentName');

if(!($id eq null) && $experimentName eq 'pilot') {
	my $repeat = 0;

	# Open participants file to check time limit 
	my $statusFile = '/var/www/html/TongLab/mTurk/pilot-jspsych-experiment/conditions/status.txt';
	my $partFile = '/var/www/html/TongLab/mTurk/pilot-jspsych-experiment/conditions/participants.txt';

	my $json_stext = do {
	   open(my $json_sf, "<:encoding(UTF-8)", $statusFile)
	      or die("Can't open \$statusFile\": $!\n");
	   local $/;
	   <$json_sf>
	};

	my @status = @{decode_json $json_stext};

	my $json_ptext = do {
	   open(my $json_pf, "<:encoding(UTF-8)", $partFile)
	      or die("Can't open \$partFile\": $!\n");
	   local $/;
	   <$json_pf>
	};

	my @participants = @{decode_json $json_ptext};

	foreach $slot (0..$#participants) {
	  if(@{@participants[$slot]}[0] eq $id) {
	  	$repeat = 1;
	  	last;
	  }
	}
	print $id;
	print @{@participants[$index]}[0];
	print $repeat;
	if(!$repeat) {
		$place;
		foreach $index (0..$#status) {
		  if($status[$index]==0) {
		  	$place = $index;
		  	$status[$place] = 1;
		  	last;
		  }
		}

		@{@participants[$place]}[0] = $id;
		@{@participants[$place]}[1] = time();

		my $json_str_status = encode_json \@status;
		my $json_str_participants = encode_json \@participants;

		open(my $SID, '>', $statusFile) 
			or die("Can't open \$statusFile\": $!\n");
		print $SID "$json_str_status";

		open(my $PID, '>', $partFile) 
			or die("Can't open \$partFile\": $!\n");
		print $PID "$json_str_participants";

		close(status);
		close(participants);
	}
}